drivers/adc/adc_ads1119.c

```diff
@@ -0,0 +1,519 @@
+		if (channel_cfg->input_positive == 0 && channel_cfg->input_negative == 1) {
+			config |= ADS1119_CONFIG(MUX, ADS1119_CONFIG_MUX_DIFF_0_1);
+		} else if (channel_cfg->input_positive == 1 && channel_cfg->input_negative == 2) {
+			config |= ADS1119_CONFIG(MUX, ADS1119_CONFIG_MUX_DIFF_1_2);
+		} else if (channel_cfg->input_positive == 2 && channel_cfg->input_negative == 3) {
+			config |= ADS1119_CONFIG(MUX, ADS1119_CONFIG_MUX_DIFF_2_3);
+		} else {
+			return -ENOTSUP;
+		}
+	} else {
+		if (channel_cfg->input_positive == 0) {
+			config |= ADS1119_CONFIG(MUX, ADS1119_CONFIG_MUX_SINGLE_0);
+		} else if (channel_cfg->input_positive == 1) {
+			config |= ADS1119_CONFIG(MUX, ADS1119_CONFIG_MUX_SINGLE_1);
+		} else if (channel_cfg->input_positive == 2) {
+			config |= ADS1119_CONFIG(MUX, ADS1119_CONFIG_MUX_SINGLE_2);
+		} else if (channel_cfg->input_positive == 3) {
+			config |= ADS1119_CONFIG(MUX, ADS1119_CONFIG_MUX_SINGLE_3);
+		} else {
+			return -ENOTSUP;
+		}
+	}
+
+	dr = ads1119_acq_time_to_dr(dev, channel_cfg->acquisition_time);
+	if (dr < 0) {
+		return -ENOTSUP;
+	}
+
+	config |= ADS1119_CONFIG(DR, dr);
+
+	switch (channel_cfg->gain) {
+	case ADC_GAIN_1:
+		config |= ADS1119_CONFIG(GAIN, ADS1119_CONFIG_GAIN_1);
+		break;
+	case ADC_GAIN_4:
+		config |= ADS1119_CONFIG(GAIN, ADS1119_CONFIG_GAIN_4);
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	config |= ADS1119_CONFIG(CM, ADS1119_CONFIG_CM_SINGLE); /* Only single shot supported */
+
+	return ads1119_write_reg(dev, config);
+}
+
+
+static int ads1119_validate_buffer_size(const struct adc_sequence *sequence)
+{
+	size_t needed = sizeof(int16_t);
+
+	if (sequence->options) {
+		needed *= (1 + sequence->options->extra_samplings);
+	}
+
+	if (sequence->buffer_size < needed) {
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int ads1119_validate_sequence(const struct adc_sequence *sequence)
+{
+	if (sequence->resolution != ADS1119_RESOLUTION) {
+		return -ENOTSUP;
+	}
+
+	if (sequence->channels != BIT(0)) {
+		return -ENOTSUP;
+	}
+
+	if (sequence->oversampling) {
+		return -ENOTSUP;
+	}
+
+	return ads1119_validate_buffer_size(sequence);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx,
+					      bool repeat_sampling)
+{
+	struct ads1119_data *data = CONTAINER_OF(ctx,
+						 struct ads1119_data,
+						 ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->buffer_ptr;
+	}
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct ads1119_data *data = CONTAINER_OF(ctx,
+						 struct ads1119_data, ctx);
+
+	data->buffer_ptr = data->buffer;
+	k_sem_give(&data->acq_sem);
+}
+
+static int ads1119_adc_start_read(const struct device *dev,
+				  const struct adc_sequence *sequence,
+				  bool wait)
+{
+	int rc;
+	struct ads1119_data *data = dev->data;
+
+	rc = ads1119_validate_sequence(sequence);
+	if (rc != 0) {
+		return rc;
+	}
+
+	data->buffer = sequence->buffer;
+
+	adc_context_start_read(&data->ctx, sequence);
+
+	if (wait) {
```

Comment:
I'm still not convinced `wait` is needed here, after experimenting myself

https://github.com/zephyrproject-rtos/zephyr/blob/64af112d30cd7340f2a32a4eaebe86cd995bdd06/drivers/adc/adc_ads1x1x.c#L511

Reply:
The reason why this is working in your driver is because the additional thread is created even if CONFIG_ADC_ASYNC` is set to `n` which is not the case in this one. When a single thread is used we need to make that the sync semaphore is only taken if already given